server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala

```diff
@@ -0,0 +1,240 @@
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.{blocking, Await, ExecutionContext, Future}
+import scala.concurrent.duration._
+import scala.language.postfixOps
+import scala.util.{Failure, Success, Try}
+
+import org.apache.hadoop.yarn.api.records.{ApplicationId, ApplicationReport, FinalApplicationStatus, YarnApplicationState}
+import org.apache.hadoop.yarn.client.api.YarnClient
+import org.apache.hadoop.yarn.conf.YarnConfiguration
+
+import com.cloudera.livy.Logging
+import com.cloudera.livy.util.LineBufferedProcess
+
+object SparkYarnApp extends Logging {
+  private class NonDaemonThreadFactory extends ThreadFactory {
+    def newThread(r: Runnable): Thread = {
+      val t = new Thread(r)
+      t.setDaemon(true)
+      t
+    }
+  }
+
+  // Create a non daemon thread pool to run getAppIdFromTagAsync().
+  // getAppIdFromTagAsync() might take a while and should not use the global default thread pool.
+  private implicit val ec = ExecutionContext.fromExecutor(
+    Executors.newCachedThreadPool(new NonDaemonThreadFactory()))
+
+  // It takes at least 5 seconds to see the newly submitted app.
+  private val APP_TAG_TO_ID_TIMEOUT = 30 seconds
+  private val KILL_TIMEOUT = APP_TAG_TO_ID_TIMEOUT
+  private val POLL_INTERVAL = 1 second
+
+  // YarnClient is thread safe. Create once, share it across threads.
+  private[this] lazy val yarnClient = YarnClient.createYarnClient()
+  private[this] var yarnClientCreated = false
+
+  def getYarnClient(): YarnClient = synchronized {
+    if (!yarnClientCreated) {
+      yarnClient.init(new YarnConfiguration())
+      yarnClient.start()
+      yarnClientCreated = true
+    }
+    yarnClient
+  }
+
+  def getAppIdFromTagAsync(appTag: String): Future[ApplicationId] =
+    Future { getAppIdFromTag(appTag).get }
+
+  /**
+   * Find the corresponding YARN application id from an application tag.
+   *
+   * @param appTag The application tag tagged on the target application.
+   *               If the tag is not unique, it returns the first application it found.
+   *               It will be converted to lower case to match YARN's behaviour.
+   * @return ApplicationId or the failure.
+   */
+  @tailrec
+  private def getAppIdFromTag(
+    appTag: String,
+    deadline: Deadline = APP_TAG_TO_ID_TIMEOUT.fromNow): Try[ApplicationId] = {
+    val appTagLowerCase = appTag.toLowerCase
+
+    // FIXME Should not loop thru all YARN applications but YarnClient doesn't offer an API.
+    // Consider calling rmClient in YarnClient directly.
+    getYarnClient().getApplications().asScala.find(_.getApplicationTags.contains(appTagLowerCase))
+      match {
+        case Some(app) => Success(app.getApplicationId)
+        case None =>
+          if (deadline.isOverdue) {
+            Failure(new Exception(s"No YARN application is tagged with $appTagLowerCase."))
+          } else {
+            blocking { Clock.sleep(POLL_INTERVAL.toMillis) }
```

Comment:
I generally dislike code that blocks threads in a thread pool. That can be replaced by:

- using a `Promise` to signal the completion of the task to the caller (`getAppIdFromTagAsync` can still return a future)
- using a `ScheduledExecutorService` instead of a plain one, and re-scheduling a task instead of sleeping